cli/azd/test/functional/testdata/samples/containerapp/infra/main.parameters.json

```diff
@@ -8,8 +8,8 @@
     "location": {
       "value": "${AZURE_LOCATION}"
     },
-    "provisionContainerApp": {
-      "value": "${AZURE_PROVISION_CONTAINER_APP}"
+    "imageName": {
+      "value": "${SERVICE_WEB_IMAGE_NAME:=$(containerAppImage web main)}"
```

Comment:
Thinking about this logically. Why do we allow `imageName` as a parameter other than to roundtrip the value? And if so, is the design flawed? I also don't feel good that we have to ask users to explicitly opt-in to this model to gain normalcy.

An existence check in `bicep` would maybe be more palatable. Still, we've created a design language that is hard for users to understand.

Reply:
> Thinking about this logically. Why do we allow imageName as a parameter other than to roundtrip the value? 

If we look at how our container app story has evolved, we have always been in a position where the container app resource itself was initially provisioned inside the IaC file for the application.  We did this because we believe in an infrastructure as code approach and being able to leverage the existing IaC provider to create the resource seemed like the right thing to do.

A problem we found ourselves with is that when creating the container app resource, we must provide the image we want to run.  This image has to exist and be pullable by the resource in order for ARM to provision it.  If it is not there, the deployment fails, and you can't create the container app.

For an initial `provision` we don't yet have the image we intend to run, because the initial provision is going to create a container registry that we'll push the image to later.  Even if we've built the docker image locally, we can't make it available to the container apps service without pushing it to a remote registry and that registry won't exist until we've done our initial provision.

Furthermore, our strategy for `deploy` needs the container app to exist, because it works by taking the current settings, and creating a new copy that references the just built image.  

Because the container app needs to exist, it means the initial provision has to create it, and so we use a dummy image like `nginx:latest` or the Container App Hello World image, with the expectation that it will be replaced later when `azd deploy` is run.

Due to how the ARM Deployment model works, if we don't use the same image name as what is currently in use, the ARM will attempt to update the container app resource to use whatever image name we provided in the template.  This means that out of the box, running a provision then a deploy and then a provision again would have the possibility of resetting the container image back to its initial value, unless the second provision took care to provide the same image name that `azd deploy` had used.

Faced with that problem, we leveraged the fact that `SERVICE_WEB_IMAGE_NAME` was written into the environment and we could use environment variables in our `.parameters.json` files get to a place where you can pass in the image to use and then modified our bicep to trust this value if it was set and if it was unset just use the default image.

Adding the `IMAGE_NAME` stuff to the environment at all was an artifact of an earlier iteration of the container apps deployment story (where `azd deploy` would actually do an ARM deployment instead of talking to the container apps management plane directly).  In that case it was here's the image ID of the just built image, go do a bicep deployment to update the container app.

With this new command, it may be better to just do this:

```json
"imageName": {
  "value": "$(containerAppImage web main)"
}
```

And not worry about SERVICE_WEB_IMAGE_NAME at all.  We either use the live image (if the resource exists, this is the subsequent provision case) or if it fails this command returns the empty string and the bicep knows this is the initial provision case and can use a dummy image.

> And if so, is the design flawed? I also don't feel good that we have to ask users to explicitly opt-in to this model to gain normalcy.

I do think the model of doing the provision of the container apps at all in Bicep is leading us to do a bunch of weird gymnastics.  I'm hopeful that we can get out of it by adopting the manifest based deployment approach and I'm happy we are prototyping things in that direction.  I do think that it prevents this sort of issue "by construction" which is very nice.

Having said that, I do think there's value in solving the end-to-end problem on the current system if we can.

> An existence check in bicep would maybe be more palatable. Still, we've created a design language that is hard for users to understand.

In an earlier iteration I actually had this command be called `serviceResourceExists` and it took the name of the service, did a `GetTargetResource` call on it and evaluated to either "true" or "false".  When I went to go use that in the bicep for this integration test, I realized that the only thing I was going to do was pull the image name from the existing resource and flow it here.  So I took a moment and thought "if I do a command line the above, then I can save myself the trouble of having to go edit existing bicep files, because all our container apps samples follow this pattern of passing in the image to use.

It also addressed some of the feedback Jon had in #2052 around the cost of adopting the approach outlined there.